x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -82,48 +89,151 @@ func (s *websocketStream) FollowStream(ctx context.Context) error {
+			_, message, err := c.ReadMessage()
+			if err != nil {
+				s.metrics.errorsTotal.Inc()
+				if isRetryableError(err) {
+					s.log.Debugw("websocket connection encountered an error, attempting to reconnect...", "error", err)
+					// close the old connection and reconnect
+					c.Close()
+					// since c is already a pointer, we can reassign it to the new connection and the defer will still handle it
+					c, resp, err = connectWebSocket(ctx, s.cfg, url, s.log)
+					handleConnectionResponse(resp, s.log)
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("failed to reconnect websocket connection", "error", err)
+						return err
+					}
+				} else {
+					s.log.Errorw("failed to read websocket data", "error", err)
+					return err
+				}
+			}
+			s.metrics.receivedBytesTotal.Add(uint64(len(message)))
+			state["response"] = message
+			s.log.Debugw("received websocket message", logp.Namespace("websocket"), string(message))
+			err = s.process(ctx, state, s.cursor, s.now().In(time.UTC))
+			if err != nil {
+				s.metrics.errorsTotal.Inc()
+				s.log.Errorw("failed to process and publish data", "error", err)
+				return err
+			}
+		}
+	}
+}
+
+// isRetryableError checks if the error is retryable based on the error type.
+func isRetryableError(err error) bool {
+	// check for specific network errors
+	var netErr *net.OpError
+	if errors.As(err, &netErr) {
+		switch {
+		case netErr.Op == "dial" && netErr.Err.Error() == "i/o timeout",
+			netErr.Op == "read" && netErr.Err.Error() == "i/o timeout",
+			netErr.Op == "read" && netErr.Err.Error() == "connection reset by peer",
+			netErr.Op == "read" && netErr.Err.Error() == "connection refused",
+			netErr.Op == "read" && netErr.Err.Error() == "connection reset",
+			netErr.Op == "read" && netErr.Err.Error() == "connection closed":
+			return true
+		}
+	}
+
+	// check for specific websocket close errors
+	var closeErr *websocket.CloseError
+	if errors.As(err, &closeErr) {
+		switch closeErr.Code {
+		case websocket.CloseGoingAway,
+			websocket.CloseNormalClosure,
+			websocket.CloseInternalServerErr,
+			websocket.CloseTryAgainLater,
+			websocket.CloseServiceRestart,
+			websocket.CloseTLSHandshake:
+			return true
+		}
+	}
+
+	// check for common error patterns
+	if strings.Contains(err.Error(), "timeout") ||
+		strings.Contains(err.Error(), "connection reset") ||
+		strings.Contains(err.Error(), "temporary failure") ||
+		strings.Contains(err.Error(), "server is busy") {
+		return true
+	}
+
+	return false
+}
+
+// handleConnectionResponse logs the response body of the websocket connection.
+func handleConnectionResponse(resp *http.Response, log *logp.Logger) {
 	if resp != nil && resp.Body != nil {
 		var buf bytes.Buffer
-		if s.log.Core().Enabled(zapcore.DebugLevel) {
+		if log.Core().Enabled(zapcore.DebugLevel) {
 			const limit = 1e4
+			//nolint:errcheck // ignore error since if this fails it signals deeper issues with the system and the code should panic
```

Comment:
Ideally, Filebeat would never panic because of an issue from a single input, once the startup code has run, it should never panic.

I know part of this code was already there, but `handleConnectionResponse` is called by a method that returns an error, so I believe it is better to check for this error and return it wrapped with more context information, `FollowStream` can then return the error and the input can stop, if necessary.

This will stop the input but prevent Filebeat from crashing and stopping inputs that can keep working.

Reply:
**handleConnection** response is only responsible for debug logging the websocket response, in this scenario I feel it's best to just log the error as only **io.CopyN** which is a system level function, can cause an error. Stopping the websocket functionality for a copy error for debug logging might be counterintuitive. So yes I agree with handling the error and logging it but don't feel we should return it and cause the input to stop or restart.

Reply:
Suggest appending the error message parenthetically to `buf` in the case that `err != nil`.
```
if err != nil {
    fmt.Fprintf(" (%s) ", err)
}
```